train_model_lightning.py

```diff
@@ -0,0 +1,188 @@
+torch.backends.cuda.matmul.allow_tf32 = True
+
+class CustomCheckpoint(ModelCheckpoint):
+    def __init__(self, dirpath, filename, save_top_k, monitor, mode):
+        super().__init__(dirpath=dirpath, filename=filename, save_top_k=save_top_k, monitor=monitor, mode=mode)
+        self.num_ckpts = 0
+    
+    def on_save_checkpoint(self, trainer, pl_module, checkpoint):
+        super().on_save_checkpoint(trainer=trainer, pl_module=pl_module, checkpoint=checkpoint)
+        pl_module.save_pretrained(os.path.join(self.dirpath, f"hf_ckpt_{self.num_ckpts}"))
+        self.num_ckpts += 1
+
+def train_model(config: Struct):
+    arg_dict = locals()
+
+    # Test that the head dimension will be an even, whole number
+    assert config.embed_dim % (config.heads * 2) == 0, \
+        "Head Dimension must be even to perform Rotary Position Embedding " + \
+        f"({config.embed_dim} / {config.heads} = {config.embed_dim / config.heads} " + \
+        "-- not an even, whole number)! Try changing the Embedding " + \
+        "Dimension or number of heads."
+
+    # Test that the value embedding dimension is divisible by number of heads
+    assert config.value_embed_dim % config.heads == 0, \
+        "Value Embed Dimension not divisible by number of heads " + \
+        f"({config.value_embed_dim} % {config.heads} != 0)!"
+
+    # Set random seeds for torch, numpy, random, etc. with transformers library
+    if config.rand_seed is not None:
+        set_seed(config.rand_seed)
+
+    # Create requested model
+    if config.model_type.lower() == "retnet":
+        model = RetNetModel(config)
+    elif config.model_type.lower() == "transformer":
+        model = TransformerModel(config)
+    else:
+        raise ValueError(f"Model type '{config.model_type}' not supported!")
+
+    # Print all arguments for recordkeeping
+    print("Arguments:")
+    arg_table = []
+    row = []
+    for i, arg in enumerate(arg_dict.keys()):
+        row.append(f"{arg}: {arg_dict[arg]}")
+        if (i + 1) % 4 == 0:
+            arg_table.append(row)
+            row = []
+    if row:
+        arg_table.append(row)
+    print(tabulate(arg_table, tablefmt="grid"))
+
+    # Print model info
+    print("\nModel Summary:")
+    total_params = model_summary(
+        model,
+        input_data=torch.ones(1, config.seq_len).long()).total_params
+    
+    # Create unique label for model (model type, parameter count, hyperparameters**)
+    model_label = f"{config.model_type}_{total_params}_LR{config.learning_rate}_ED{config.embed_dim}_FFN{config.ffn_dim}_H{config.heads}_S{config.seq_len}_{datetime.now().strftime('%Y-%m-%d-%H:%M:%S')}"
+
+    # Initialize model directory for config files, weights, etc.
+    model_dir = Path(config.models_path) / model_label
+    model_dir.mkdir(parents=True, exist_ok=False)
+    print(f"Saving model files in {model_dir}")
+
+    # Initialize checkpoints directory
+    checkpoints_dir = model_dir / "checkpoints"
```

Comment:
Seems like this should be refactored to use the new config parameter.

Reply:
See other comment on config parameter